File: Examples/sample.go
// Prism Go sample.
//
// Exercises:
//   - package, import (grouped)
//   - struct / interface
//   - methods with pointer/value receivers
//   - goroutines, channels, select
//   - raw strings (`...`), interpolation via fmt
//   - rune literals ('x')
//   - generics (Go 1.18+)

package main

import (
	"context"
	"errors"
	"fmt"
	"log"
	"sort"
	"sync"
	"time"
)

// -----------------------------------------------------------------------------
// Constants — every literal form Go supports.
// -----------------------------------------------------------------------------

const (
	PI         float64 = 3.14159_26535
	HexMask    uint32  = 0xDEAD_BEEF
	Octal      int     = 0o755
	Binary     int     = 0b1010_1010
	Scientific float64 = 6.022e23
	NewlineCh  rune    = '\n'
	UnicodeCh  rune    = 'é'
)

// -----------------------------------------------------------------------------
// Types
// -----------------------------------------------------------------------------

type Shape interface {
	Area() float64
	Kind() string
}

type Circle struct {
	Radius float64
}

func (c Circle) Area() float64 { return PI * c.Radius * c.Radius }
func (c Circle) Kind() string  { return "circle" }

type Rect struct {
	Width, Height float64
}

func (r *Rect) Area() float64 { return r.Width * r.Height }
func (r *Rect) Kind() string  { return "rect" }

// Generic constraint
type Numeric interface {
	~int | ~int64 | ~float32 | ~float64
}

func SumOfSquares[T Numeric](values []T) T {
	var total T
	for _, v := range values {
		total += v * v
	}
	return total
}

// -----------------------------------------------------------------------------
// Worker pool with channels + select
// -----------------------------------------------------------------------------

func computeSquares(ctx context.Context, in <-chan int, out chan<- int, wg *sync.WaitGroup) {
	defer wg.Done()
	for {
		select {
		case <-ctx.Done():
			return
		case v, ok := <-in:
			if !ok {
				return
			}
			out <- v * v
		}
	}
}

// -----------------------------------------------------------------------------
// main
// -----------------------------------------------------------------------------

func main() {
	banner := `
Prism Go sample
===============
version : 1.0
mask    : 0xDEADBEEF
`
	fmt.Print(banner)

	shapes := []Shape{
		Circle{Radius: 2.0},
		&Rect{Width: 3.0, Height: 4.0},
	}
	sort.SliceStable(shapes, func(i, j int) bool { return shapes[i].Area() < shapes[j].Area() })

	for _, s := range shapes {
		fmt.Printf("%s: area=%.3f\n", s.Kind(), s.Area())
	}

	fmt.Printf("sum of squares: %d\n", SumOfSquares([]int{1, 2, 3, 4, 5}))

	// Goroutines + channels
	ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
	defer cancel()

	in := make(chan int)
	out := make(chan int)
	var wg sync.WaitGroup
	wg.Add(2)
	go computeSquares(ctx, in, out, &wg)
	go computeSquares(ctx, in, out, &wg)

	go func() {
		defer close(in)
		for i := 1; i <= 5; i++ {
			in <- i
		}
	}()

	go func() {
		wg.Wait()
		close(out)
	}()

	results := make([]int, 0, 5)
	for v := range out {
		results = append(results, v)
	}
	fmt.Println("squares =", results)

	if err := mightFail(); err != nil {
		log.Printf("warning: %v\n", err)
	}
}

func mightFail() error {
	return fmt.Errorf("not really an error: %w", errors.New("stub"))
}

File: Examples/sample.rs
// Prism Rust sample.
//
// Exercises:
//   - line, /// doc, /** ... */, and nested /* /* */ */ block comments
//   - raw strings r"..." / r#"..."#  and byte strings b"..."
//   - char vs lifetime: 'x' (char) vs 'a (lifetime)
//   - macros (println!, vec!, format!) and #[attr] / #![attr]
//   - traits, impl blocks, generics with bounds
//   - pattern matching, Option/Result, ? operator
//   - async (without runtime — just syntactic showcase)

#![allow(dead_code, unused_variables)]
#![cfg_attr(test, allow(unused_imports))]

use std::collections::HashMap;
use std::fmt::{self, Display, Formatter};

/// Crate-level constants — every numeric form.
const PI: f64        = 3.14159_26535_f64;
const HEX_MASK: u32  = 0xDEAD_BEEF;
const OCTAL: u32     = 0o755;
const BINARY: u8     = 0b1010_1010;
const SCIENTIFIC: f64 = 6.022e23;
const NEWLINE: char  = '\n';
const UNICODE: char  = 'é';

/* -----------------------------------------------------------------------------
 * Trait + generic data type
 * -- nested /* like /* this */ */ -- block comment showcase
 * ----------------------------------------------------------------------------- */

trait Shape {
    fn area(&self) -> f64;
    fn kind(&self) -> &'static str;
}

#[derive(Debug, Clone, Copy, PartialEq)]
struct Circle { radius: f64 }

impl Shape for Circle {
    fn area(&self) -> f64 { PI * self.radius * self.radius }
    fn kind(&self) -> &'static str { "circle" }
}

#[derive(Debug, Clone, Copy, PartialEq)]
struct Rect { width: f64, height: f64 }

impl Shape for Rect {
    fn area(&self) -> f64 { self.width * self.height }
    fn kind(&self) -> &'static str { "rect" }
}

/// Demonstrates lifetimes and trait bounds.
fn largest<'a, T: Shape>(items: &'a [T]) -> Option<&'a T> {
    items.iter().reduce(|a, b| if a.area() >= b.area() { a } else { b })
}

/* -----------------------------------------------------------------------------
 * Enum + pattern matching
 * ----------------------------------------------------------------------------- */

enum Severity {
    Debug,
    Info,
    Warn,
    Error(String),
}

impl Display for Severity {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        match self {
            Severity::Debug      => write!(f, "DEBUG"),
            Severity::Info       => write!(f, "INFO"),
            Severity::Warn       => write!(f, "WARN"),
            Severity::Error(msg) => write!(f, "ERROR: {msg}"),
        }
    }
}

/* -----------------------------------------------------------------------------
 * Strings: regular, raw, byte
 * ----------------------------------------------------------------------------- */

fn strings_showcase() -> String {
    let plain   = "hello, world";
    let escaped = "tab\there, newline\n";
    let raw     = r"contains a literal backslash: \d+ and \n stays literal";
    let raw_h   = r#"a "raw" string with embedded quotes"#;
    let bytes   = b"raw\xDEbytes";
    format!("{plain} :: {escaped} :: {raw} :: {raw_h} :: bytes={:?}", bytes)
}

/* -----------------------------------------------------------------------------
 * Async + Result
 * ----------------------------------------------------------------------------- */

async fn fetch_square(value: i32) -> Result<i32, &'static str> {
    if value < 0 { Err("negative") } else { Ok(value * value) }
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("Prism Rust sample — mask={HEX_MASK:#X}");

    let shapes: Vec<Box<dyn Shape>> = vec![
        Box::new(Circle { radius: 2.0 }),
        Box::new(Rect   { width: 3.0, height: 4.0 }),
    ];
    for s in &shapes {
        println!("{}: area={:.3}", s.kind(), s.area());
    }

    let circles = [Circle { radius: 1.0 }, Circle { radius: 5.0 }];
    if let Some(c) = largest(&circles) {
        println!("largest circle: {c:?}");
    }

    let mut levels: HashMap<&str, Severity> = HashMap::new();
    levels.insert("default", Severity::Info);
    levels.insert("io", Severity::Error("disk full".into()));

    for (component, level) in &levels {
        println!("[{component}] {level}");
    }

    println!("{}", strings_showcase());

    let squares: Vec<i32> = (1..=5)
        .map(|n| futures::executor::block_on(fetch_square(n)).unwrap())
        .collect();
    println!("squares = {squares:?}");

    Ok(())
}
